lib/widgets/message_list.dart

```diff
@@ -247,6 +244,30 @@ class _MessageListState extends State<MessageList> with PerAccountStoreAwareStat
   Widget _buildListView(context) {
     final length = model!.items.length;
     return StickyHeaderListView.builder(
+      findChildIndexCallback: (Key key) {
+        // To preserve state across rebuilds for individual [MessageItem]
+        // widgets as the size of [MessageListView.items] changes we need
+        // to match old widgets by their key to their new position in
+        // the list.
+        //
+        // The keys are of type [ValueKey] with a value of [Message.id]
+        // and here we use a O(log n) binary search method. This could
+        // be improved but for now it only triggers for materialized
+        // widgets. As a simple test, flinging through All Messages in
+        // CZO on a Pixel 5, this only runs about 10 times per rebuild
+        // and the timing for each call is <100 microseconds.
+        //
+        // Non-message items (e.g., start and end markers) that do not
+        // have state that needs to be preserved have not been given keys
+        // and will not trigger this callback.
+        final valueKey = key as ValueKey;
+        final index = model!.findItemWithMessageId(valueKey.value);
+        if (index == -1) {
+          return null;
+        }
+        // Note `itemBuilder` pulls from `model.items` in reverse order.
+        return length - 1 - index;
```

Comment:
```suggestion
        return length - 1 - index;
```

The corresponding code in `itemBuilder` isn't far away, and that's the most dependable thing to refer to to understand why this expression is what it is.

Also we should have tests that check this is correct. That's the dependable way to make sure that this and `itemBuilder` stay in sync. … And at a quick experiment, things look good on that front — if I edit this to `return index;`, several tests fail.

Reply:
> The corresponding code in `itemBuilder` isn't far away

Speaking of which, though, a nit: this `findChildIndexCallback` should go after `keyboardDismissBehavior`, not before. That way it's adjacent to the stanza that has `itemBuilder` in it, which is good because as we see their logic is interdependent with each other. And there isn't such an interdependence involving `keyboardDismissBehavior`.